## How to run AmpliSolve ?? 

First make sure that you have the correct environment, resolved all programâ€™s dependencies and generated the binaries: 

```
AmpliSolveErrorEstimation 
```
and 

```
AmpliSolveVariantCalling
```

### Prepare the input files for AmpliSolve

AmpliSolve requires a pre-processing step that utilizes ASEQ. To make the execution faster, this is a separate step that can be executed in parallel using a bash script for example in a cluster. ASEQ software is available at https://demichelislab.unitn.it/doku.php?id=public:aseq , but we also provide a simplified version of ASEQ named computeCounts (only the binary, please contact us for more info) in our Versions_used_during_development DIR. This program deploys the PILEUP mode of ASEQ which is the one we need for this pre-processing step. 

We recommend to place all available normal samples (germline) in a separete DIR say NORMAL_BAM_DIR , and all tumour samples in another DIR named TUMOUR_BAM_DIR 

To run ASEQ (or equally computeCounts) type: 

```
./ASEQ [vcf=dummyVCF.txt] [bam=myBAM.bam] [threads=int] [mbq=int] [mrq=int] [mdc=int] [out=Out_DIR]
```

where: 
1. dummyVCF.txt is a VCF-like file that contains all positions in the gene panel. Remember that this is not the actual panel design, as the panel design contains amplicons written as intervals, whereas here we enumate all positions in the panel. 

Assuming that we have a small panel with one amplicon chr8:23437012-23437020 the dummyVCF.txt will look like:

```
chr8	23437012	.	.	.	.	.	.
chr8	23437013	.	.	.	.	.	.
chr8	23437014	.	.	.	.	.	.
chr8	23437015	.	.	.	.	.	.
chr8	23437016	.	.	.	.	.	.
chr8	23437017	.	.	.	.	.	.
chr8	23437018	.	.	.	.	.	.
chr8	23437019	.	.	.	.	.	.
chr8	23437020	.	.	.	.	.	.
```
2. myBam.bam is the bam files of interest. Remember that we need to run this command for ALL files in NORMAL_BAM_DIR and TUMOUR_BAM_DIR 

3. threads is the number of threads to use. Usually I put 8, but this depends on your system

4. mbq, mrq , mdc are quality parameters ; we recommend the triplet 20-20-20 for Ion AmpliSeq data

5. Out_DIR is the directory to store the read count files ending with .PILEUP.ASEQ 
Since we need a DIR of normal read counts to compute the error levels, we must store all .PILEUP.ASEQ of normal samples in one DIR say NORMAL_ASEQ_DIR 

In a similar fashion, we need a DIR for all tumour read counts to perform variant calling, so we must store all .PILEUP.ASEQ of tumour samples in another DIR say TUMOUR_ASEQ_DIR 

Once we have all .PILEUP.ASEQ files for normal and tumour files stored in two separete DIRs say NORMAL_ASEQ_DIR and TUMOUR_ASEQ_DIR we are ready and we can run the AmpliSolve program. 


### Run AmpliSolveErrorEstimation

Once the input data are ready, we need to estimate the error levels per position, nucleotide and strand. This is performed using program AmpliSolveErrorEstimation. 

Type:

```
./AmpliSolveErrorEstimation panel_design=/your/panel/design/file reference_genome=/your/reference/genome germline_dir=/your/dir/with/germline/read/count/files C_value=/the/error/rescaling/factor output_dir=/dir/to/store/the/intermediate/results

```
where

1. panel_degign is a bed (chrom TAB start TAB end) file that contains all amplicons coordinates 

2. reference_genome is the FASTA format file used as reference

3. germline_dir is the DIR with the read counts generated by ASEQ (see how above). Following our previous example we give:

```
germline_dir=NORMAL_ASEQ_DIR
```

4. C_value is the normalization factor described in the manuscript. For ctDNA screening we recommend the value 0.002

5. output_dir is a DIR to store some intermediate results

#### One complete execution example looks as follows:

```
./AmpliSolveErrorEstimation panel_design=AmpliSeq_30genes_Designed-1.bed reference_genome=/Users/Reference_genome/hg19_chr.fa germline_dir=NORMAL_ASEQ_DIR C_value=0.002 output_dir=FirstExecution
```




